### ISSUE TITLE
feat(triton): llama.rs.in expressed as a llama.py using @triton.jit kernel

### ISSUE THREAD

**Issue by aavarghese** [AI-native-Systems-Research/scratchy#34]
> **In short:** `llama.rs.in` expressed as a Triton `llama.py`, then run through the
> `triton-rust` crates rather than reasoned about. Stages 1–4 (front end → TTIR → KTIR →
> tiling and corelet planning) accept the **entire** Llama forward unmodified. Every
> failure is in stage 5, KTIR → SuperDSC, and there are exactly three: the embedding
> gather, the layer loop's initial value, and `.T` over a computed tile. With attention
> removed, 38 SuperDSCs lower clean. That inverts where the five-step plan puts the risk.
> Conclusion: generate `llama.py` from `llama.rs.in` rather than replace it.
>
> The file itself: [`llama.py` in the comment below](#issuecomment-5699708837).

## What this is

Nick suggested experimenting with what `crates/models/arch/dsl/llama.rs.in` would look like
expressed instead as a `llama.py` using Triton's `@triton.jit`. This is the result.

Claude ran it through the three crates in `triton-rust.tar.gz` (`triton-frontend`, `triton-ktir`, `triton-superdsc`)

This experiment clears KTDP, gets through `plan_corelets` far enough to emit `ktdf.corelet_plan`, and never reaches unit mapping.

## The measured chain

Whole Llama forward as one `@triton.jit` kernel. `N_LAYERS=2 M=64 D_MODEL=128 D_FF=256
V=512 grid=[4]`:

| # | stage | result |
|---|-------|--------|
| 1 | `codegen` (census + `.py` → raw TTIR) | **OK** — 671 lines of TTIR |
| 2 | `opt::make_ttir` (inline/fold/cse/dce) | **OK** — 380 lines post-pass |
| 3 | `from_ttir::convert` (TTIR → KTIR) | **OK** — 166 ops in / 166 ops out |
| 4 | `make_ktir` (the seven passes, ending `plan_corelets`) | **OK** — 252 lines of KTIR |
| 5 | `triton-superdsc-lower` (KTIR → SuperDSC) | **REFUSED**, three distinct ways |

The front end and the TTIR→KTIR conversion take the entire forward pass unmodified. It
stops in the SuperDSC emitter.

## The three refusals, verbatim

**1 — the embedding gather.**

> op 68: `construct_indirect_access_tile` results in nothing, not an `!ktdp.access_tile`

Worth noting this names the kernel-tiling dialect directly. `embed()` is therefore not a
front-end hole  — it is a KTIR→SuperDSC hole. Remove the gather and read
the hidden state from a plain descriptor, and the failure moves on to:

**2 — the layer loop.** The load-bearing finding.

> op 66: loop-carried tensor `%x_5` is initialised from a value this emitter cannot seed
> (p_embed); a carried buffer's initial contents must be a compile-time splat so the
> caller contract can name them

`for layer in range(0, N_LAYERS)` compiles, converts, and reaches KTIR as a real `scf.for`
whose `iter_arg` is the hidden state and whose body is a `tt.call`. Neither construct
appears in any fixture and both work. But the emitter will only seed a carried buffer from
a compile-time splat, which is why every fixture's loop carries an accumulator starting at
`tl.zeros`. A residual stream seeded from the embedding is exactly the case it refuses.
That is the difference between `llama.rs.in`'s 22 lines and a hand-unrolled 32-layer
kernel, so it matters.

**3 — `.T` over a computed tile, i.e. K-transpose in a fused layer.**

> op 106: `tt.trans` over a COMPUTED tile ([64, 64]) has no established bundle-input form,
> and this emitter will not guess one.

The rest of that diagnostic: `interslicetranspose_fp16` is a real `opFuncName`
(`arch_enums.cpp:494`, bound to `inter_slice_transpose.ddl`, arity 1→1 at
`designSpaceConfig.cpp:9422`); it *was* emitted; `dxp_standalone -b sentient` is what
refused it — "Could not find any suitable dimension mapping,
`ddl/ddl_conversion.cpp` line 2505", all eight DimProps reporting
`PrimaryDimTypes = Invalid`, `DropDim = T`, `dimCandidates = []`. It is a shape gap: every
template the crate binds today has empty `slice_layout`/`stick_layout`, while
`inter_slice_transpose.ddl` wants an eight-dim space.

And explicitly *not* attention's `.T` from the fixtures: there the operand comes from
caller-owned memory and folds into `access_tile_order` as a host obligation, whereas this
one is a register tile the host cannot present. The distinction is structural, not
stylistic — in a **fused** layer K comes out of a projection, so K-transpose is always
computed. The fixtures avoid it only by taking K from a descriptor.

## What does lower all the way

Cut attention, keep rmsnorm + residual + SwiGLU MLP + lm_head: stage 5 succeeds with
**38 SuperDSCs** plus `bundle.mlir` and `manifest.json`. Three notes from that run are
facts about the target rather than about this file:

- `arith.negf` + `math.exp` + `arith.addf` + `arith.divf` is pattern-matched into a single
  `silu` SuperDSC (`OpFuncs::SILU_FWD` → `unary_parallel.ddl:29`, no ISA gate). The note
  adds that the tolerance derived in `test/pod/swiglu_eps_derivation.py` applies to the
  **fused** form only — "the unfused four-op form would be a DIFFERENT algorithm and that
  tolerance would not".
- the `D_FF` `scf.for` carries nothing across 8 trips and is **unrolled**, because the
  flattened tile id it induces is delinearised with `arith.remui`/`arith.divui` — not
  affine, so no rolled address expression exists.
- the contraction's accumulator is materialised as a separate `add` SuperDSC rather than
  `bmm.ddl`'s `genericpartialreduction` epilogue.

## Two constructs that had to change to pass stage 1

- `.gather` needs a **one-row** descriptor block (`[1, D_MODEL]`, not `[M, D_MODEL]`) —
  the gathered row count comes from the index vector, so a multi-row block has no meaning.
- **every** `tl.dot` needs `out_dtype=tl.float16`, and a running sum needs `acc=`. The
  default is f32 (`out_dtype = tl.float32 if acc is None else acc.type.element_ty`,
  `semantic.rs:1644`), so `tl.dot(p, v)` where `v` itself came from a dot is a dtype
  mismatch. All `decoder_block.py` goldens are `tt.dot %a, %b, %acc -> f16`.

## The risk inversion

The five-step Rust beachhead treats Triton.py→TTIR and TTIR→KTIR as the unbuilt, risky
work and KTIR→DSC as more or less done. This run says the opposite: the top two stages
swallowed the entire Llama forward unmodified, and all three failures are in the bottom
one.

The three also share a shape. SuperDSC's caller contract wants a host-presentable operand,
a compile-time-nameable buffer seed, and an affine address expression — three facts the
fused form structurally cannot present. That looks like the "we never go back uphill"
thesis showing up as three concrete error messages.

## What this means for `llama.rs.in`

`llama.py` is 222 lines of Triton against `llama.rs.in`'s 22, and nearly all the delta is
bookkeeping: 17 `make_tensor_descriptor` calls over 36 lines, and one `tt.call` whose 12
descriptors flatten to **57 SSA operands** under a ~340-character mangled name. The layer
loop is only possible because each weight kind is stacked into one
`[N_LAYERS*rows, cols]` descriptor and indexed at `[layer*D_MODEL, 0]`, so the loop
carries only `x`; unrolled, 32 layers would be ~384 kernel arguments.

So the conclusion to draw is that the interesting version is **not** `llama.py`
instead of `llama.rs.in` — it is scratchy **emitting** `llama.py`. That buys "every
construct is standard Triton, defined by Triton's own toolchain" without anyone
hand-writing 250 lines per model, and keeps the 22-line file as the thing a human touches.

## Still not expressible at any stage

- `rope_append`'s append half. Writing K/V into a paged cache is a **scatter**; the only
  address-is-data op in the census is `gather`, which is a load.
- paged attention over a block table — same reason. (The goldens' `loc` paths come from a
  worktree named `paged-attn-lowering`, so this is presumably in flight.)
- multi-head. There is no concatenate, and `tl.reshape`/`tl.permute`/`tl.split` are all
  refused, so per-head outputs cannot be joined into `[M, D_MODEL]` before `o_proj`. This
  is why `decoder_block.py` is one head.

## Reproducing this

`llama.py`, the three integration tests, and the emitted TTIR/KTIR are **not pushed
anywhere yet** — they are local, deliberately outside this repo since the tarball crates
are not in it either.

Two build findings from getting the chain to run at all:

- `triton-ktir/Cargo.toml:43` pins `ktir-core` to
  `/Users/nickm/git/scratchy/.claude/worktrees/ktir-retarget/...`, so that crate builds on
  one machine only. Claude worked around it by fetching `origin/ktir-retarget` and vendoring
  `crates/targets/spyre/ktir` via `git archive` (no change to any working tree), then
  repointing that one line at the vendored copy.
- `cargo --offline` cannot resolve `ruff_python_ast` even for the dependency-free
  configuration, so the first build has to be online (84 packages).
- the `.py` fixtures are not in the tarball (they live at `../../test/fixtures`), so
  `decoder_block.py` had to be reconstructed from the `loc` metadata in
  `decoder_layer.ttir_raw.mlir`.

## One question

`decoder_block.py` pins `QK_SCALE = 0.011271055`, and `0.011271055 × 128` is exactly
`1.44269504` = `log2(e)`. So that fixture's `sm_scale` is `1/head_dim`, whereas Llama and
Granite use `1/sqrt(head_dim)` — 0.0883883 at head_dim 128, which would make `QK_SCALE`
0.1275. Deliberate arbitrary fixture constant, or is it meant to be a model's scale?

---

🤖 Investigated and written with [Claude Code](https://claude.com/claude-code) — the chain
above was run against the tarball crates, not inferred; the quoted diagnostics are the
emitter's own output.

**Comment by starpit** [comment]
Following up on my own proposal, because it was still too big. Most of those 66 lines were
restatement, and once you name the two sources of it the file collapses.

## What was actually complicated

**`__init__` restates `config.json`.** `nn.Linear(cfg.hidden_size, cfg.num_attention_heads *
cfg.head_dim, bias=False)` is arithmetic over four values we already read from the config. Every
one of those constructor calls is a second copy of a fact that already has a home.

**The class tree restates the checkpoint.** `LlamaModel` → `LlamaDecoderLayer` → `LlamaAttention`
exists so that `state_dict()` emits `model.layers.0.self_attn.q_proj.weight`. But we don't call
`state_dict()` — we read the checkpoint. The attribute path *is* the key, so declaring a class
tree to generate a key path we already have is the same duplication in a different costume.

Delete both and this is the whole model:

```python
from torch import nn

from scratchy_ops import paged_attention, rms_norm, rope_append, silu_and_mul


class LlamaForCausalLM(nn.Module):
    """crates/models/arch/dsl/llama.rs.in. Shapes come from config.json; an attribute
    path IS the checkpoint key, resolved against the real checkpoint at expansion."""

    def forward(self, input_ids, positions, rotary, kv_caches, block_table):
        x = self.model.embed_tokens(input_ids)
        for layer, kv in zip(self.model.layers, kv_caches):
            attn, mlp = layer.self_attn, layer.mlp
            h = rms_norm(x, layer.input_layernorm)
            q, k, v = attn.q_proj(h), attn.k_proj(h), attn.v_proj(h)
            q, k, v = rope_append(q, k, v, positions, rotary, kv)
            x = x + attn.o_proj(paged_attention(q, k, v, kv, block_table))
            h = rms_norm(x, layer.post_attention_layernorm)
            x = x + mlp.down_proj(silu_and_mul(mlp.gate_proj(h), mlp.up_proj(h)))
        return self.lm_head(rms_norm(x, self.model.norm))
```

**11 lines of body, 17 with imports — shorter than `llama.rs.in`'s 22**, and line-for-line it is
the same program: embed, loop, norm, qkv, rope+append, paged attention, o_proj, residual, norm,
SwiGLU, residual, final norm, lm_head. A typo in `layer.self_attn.q_porj` is a build error against
a real checkpoint, which the `.rs.in` spelling is not.

## The trade, and it is a real one

That file is **not runnable torch.** With no `__init__` you cannot instantiate it, so you cannot
`pip install torch` and run it on CPU.

If the macro *parses* rather than executes (option (a) in my previous comment — `ruff_python_ast`,
no Python in the build), runnability is a property we never consume, and then those 49 extra lines
are paying for nothing.

Except they do buy one thing, and it is the thing this repo usually insists on: **a differential
oracle.** A runnable torch model is something you can execute on CPU and diff against what the
card produced, token for token. That is worth more than its line count when a forward goes subtly
wrong — and #34's ledger is a list of exactly the subtle-wrongness this hardware produces.

## So: have both, and generate the expensive one

Let the 17-line `forward` be the source of truth — the file a human reads and edits. Then have the
macro *emit* the full `nn.Module` with a real `__init__`, synthesised from `config.json`, as a test
artifact. That is the same trick `#[forward]` already plays everywhere else: the human writes the
short declarative thing, the compile-time pipeline expands it, and the expansion is not something
anyone maintains by hand.

That gets all three: 17 lines to read, runnable torch to diff against, and zero restatement of the
config or the checkpoint — because the one copy that exists is generated from them.

It also sharpens the acceptance test from my last comment. Instead of only requiring the emitted
tape to be byte-identical to `llama.rs.in`'s, you get a second, independent check: run the
generated `nn.Module` on CPU and compare logits against the spyre bake for the same prompt. Two
oracles at different levels, neither of which moves with the thing it is checking.

## One honest note on "standard"

At 66 lines the file is real torch and an outsider can run it. At 17 lines it *reads* like torch,
but the only torch-native constructs left are `+` and calling a module — everything else is one of
our four custom ops. Simplicity and standardness pull against each other here, and the 17-line
version sits well toward the simplicity end.

I think that is still the right pick, because the legibility we actually want is "an outsider
recognises the shape of the computation and can follow it", and the short version delivers that
better than the long one. But it is a choice, not a free lunch, and it should be made deliberately
rather than by whoever writes the second model file first.

---

🤖 Written with [Claude Code](https://claude.com/claude-code).
